// ts-lib-utils/es/language-service.js
import ts from 'typescript';
import * as fs from 'fs';
/**
 * @public
 */
export function getLanguageService(uniqFiles) {
    return ts.createLanguageService({
        getCompilationSettings: function () {
            return {
                jsx: ts.JsxEmit.React
            };
        },
        getScriptFileNames: function () {
            return uniqFiles;
        },
        getScriptVersion: function () {
            return '';
        },
        getScriptSnapshot: function (fileName) {
            if (fileName === '.ts') {
                return ts.ScriptSnapshot.fromString('');
            }
            return ts.ScriptSnapshot.fromString(fs.readFileSync(fileName, { encoding: 'utf8' }));
        },
        getCurrentDirectory: function () { return '.'; },
        getDefaultLibFileName: function (options) {
            return ts.getDefaultLibFilePath(options);
        },
        fileExists: ts.sys.fileExists,
        readFile: ts.sys.readFile,
        readDirectory: ts.sys.readDirectory
    });
}
