// ts-lib-utils/es/tsconfig.js
import { __assign, __awaiter, __generator, __values } from "tslib";
import ts from 'typescript';
import * as fs from 'fs';
import * as path from 'path';
import glob from 'glob';
/**
 * @public
 */
export function getProjectRootNamesAndCompilerOptions(project) {
    return __awaiter(this, void 0, void 0, function () {
        var _a, configFilePath, dirname, config, _b, compilerOptions, errors, rootNames;
        return __generator(this, function (_c) {
            switch (_c.label) {
                case 0:
                    _a = getTsConfigFilePath(project), configFilePath = _a.configFilePath, dirname = _a.dirname;
                    config = getTsConfig(configFilePath, dirname);
                    _b = ts.convertCompilerOptionsFromJson(config.compilerOptions, config.basePath || dirname), compilerOptions = _b.options, errors = _b.errors;
                    if (errors && errors.length > 0) {
                        throw errors;
                    }
                    return [4 /*yield*/, getRootNames(config, dirname)];
                case 1:
                    rootNames = _c.sent();
                    return [2 /*return*/, { rootNames: rootNames, compilerOptions: compilerOptions }];
            }
        });
    });
}
function getTsConfigFilePath(project, fallbackProject) {
    var configFilePath;
    var dirname;
    var projectStats;
    try {
        projectStats = fs.statSync(project);
    }
    catch (error) {
        if (fallbackProject) {
            project = fallbackProject;
            projectStats = fs.statSync(project);
        }
        else {
            throw error;
        }
    }
    if (projectStats.isDirectory()) {
        configFilePath = path.resolve(project, 'tsconfig.json');
        dirname = project;
    }
    else if (projectStats.isFile()) {
        configFilePath = project;
        dirname = path.dirname(project);
    }
    else {
        throw new Error("paramter 'project' should be a file or directory.");
    }
    return { configFilePath: configFilePath, dirname: dirname };
}
function getTsConfig(configFilePath, dirname) {
    var configResult = ts.readConfigFile(configFilePath, function (p) { return fs.readFileSync(p).toString(); });
    var config = configResult.error ? {
        extends: undefined,
        compilerOptions: {
            lib: [
                'dom',
                'es5',
                'es2015',
                'es2016',
                'es2017'
            ],
            allowSyntheticDefaultImports: true
        }
    } : configResult.config;
    if (config.extends) {
        var project = void 0;
        var fallbackProject = void 0;
        if (path.isAbsolute(config.extends)) {
            project = config.extends;
        }
        else if (config.extends === '.'
            || config.extends === '..'
            || config.extends.startsWith("." + path.sep)
            || config.extends.startsWith(".." + path.sep)
            || config.extends.startsWith('./')
            || config.extends.startsWith('../')) {
            project = path.resolve(dirname, config.extends);
        }
        else {
            project = path.resolve(dirname, 'node_modules', config.extends);
            fallbackProject = path.resolve(process.cwd(), 'node_modules', config.extends);
        }
        var _a = getTsConfigFilePath(project, fallbackProject), configFilePath_1 = _a.configFilePath, extendsBasename = _a.dirname;
        var extendsConfig = getTsConfig(configFilePath_1, extendsBasename);
        config.compilerOptions = __assign(__assign({}, extendsConfig.compilerOptions), config.compilerOptions);
        config.basePath = extendsBasename;
    }
    return config;
}
function getRootNames(config, dirname) {
    return __awaiter(this, void 0, void 0, function () {
        var include, exclude, rules, include_1, include_1_1, file, currentPath, stats, e_1_1, rootNames;
        var e_1, _a;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    include = config.include;
                    exclude = config.exclude || ['node_modules/**'];
                    if (config.files) {
                        return [2 /*return*/, config.files.map(function (f) { return path.resolve(dirname, f); })];
                    }
                    if (!(include && Array.isArray(include) && include.length > 0)) return [3 /*break*/, 9];
                    rules = [];
                    _b.label = 1;
                case 1:
                    _b.trys.push([1, 6, 7, 8]);
                    include_1 = __values(include), include_1_1 = include_1.next();
                    _b.label = 2;
                case 2:
                    if (!!include_1_1.done) return [3 /*break*/, 5];
                    file = include_1_1.value;
                    currentPath = path.resolve(dirname, file);
                    return [4 /*yield*/, statAsync(currentPath)];
                case 3:
                    stats = _b.sent();
                    if (stats === undefined) {
                        rules.push(currentPath);
                    }
                    else if (stats.isDirectory()) {
                        rules.push((currentPath.endsWith('/') ? currentPath.substring(0, currentPath.length - 1) : currentPath) + "/**/*.{ts,tsx}");
                    }
                    else if (stats.isFile()) {
                        rules.push(currentPath);
                    }
                    _b.label = 4;
                case 4:
                    include_1_1 = include_1.next();
                    return [3 /*break*/, 2];
                case 5: return [3 /*break*/, 8];
                case 6:
                    e_1_1 = _b.sent();
                    e_1 = { error: e_1_1 };
                    return [3 /*break*/, 8];
                case 7:
                    try {
                        if (include_1_1 && !include_1_1.done && (_a = include_1.return)) _a.call(include_1);
                    }
                    finally { if (e_1) throw e_1.error; }
                    return [7 /*endfinally*/];
                case 8: return [2 /*return*/, globAsync(rules.length === 1 ? rules[0] : "{" + rules.join(',') + "}", exclude, dirname)];
                case 9: return [4 /*yield*/, globAsync("**/*.{ts,tsx}", exclude, dirname)];
                case 10:
                    rootNames = _b.sent();
                    return [2 /*return*/, rootNames.map(function (r) { return path.resolve(process.cwd(), dirname, r); })];
            }
        });
    });
}
function statAsync(file) {
    return new Promise(function (resolve) {
        fs.stat(file, function (error, stats) {
            if (error) {
                resolve(undefined);
            }
            else {
                resolve(stats);
            }
        });
    });
}
function globAsync(pattern, ignore, cwd) {
    return new Promise(function (resolve, reject) {
        glob(pattern, { ignore: ignore, cwd: cwd }, function (error, matches) {
            if (error) {
                reject(error);
            }
            else {
                resolve(matches);
            }
        });
    });
}
