// ts-lib-utils/dist/language-service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getLanguageService = void 0;
const tslib_1 = require("tslib");
const typescript_1 = tslib_1.__importDefault(require("typescript"));
const fs = tslib_1.__importStar(require("fs"));
/**
 * @public
 */
function getLanguageService(uniqFiles) {
    return typescript_1.default.createLanguageService({
        getCompilationSettings() {
            return {
                jsx: typescript_1.default.JsxEmit.React
            };
        },
        getScriptFileNames() {
            return uniqFiles;
        },
        getScriptVersion() {
            return '';
        },
        getScriptSnapshot(fileName) {
            if (fileName === '.ts') {
                return typescript_1.default.ScriptSnapshot.fromString('');
            }
            return typescript_1.default.ScriptSnapshot.fromString(fs.readFileSync(fileName, { encoding: 'utf8' }));
        },
        getCurrentDirectory: () => '.',
        getDefaultLibFileName(options) {
            return typescript_1.default.getDefaultLibFilePath(options);
        },
        fileExists: typescript_1.default.sys.fileExists,
        readFile: typescript_1.default.sys.readFile,
        readDirectory: typescript_1.default.sys.readDirectory
    });
}
exports.getLanguageService = getLanguageService;
