// ts-lib-utils/dist/tsconfig.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getProjectRootNamesAndCompilerOptions = void 0;
const tslib_1 = require("tslib");
const typescript_1 = tslib_1.__importDefault(require("typescript"));
const fs = tslib_1.__importStar(require("fs"));
const path = tslib_1.__importStar(require("path"));
const glob_1 = tslib_1.__importDefault(require("glob"));
/**
 * @public
 */
async function getProjectRootNamesAndCompilerOptions(project) {
    const { configFilePath, dirname } = getTsConfigFilePath(project);
    const config = getTsConfig(configFilePath, dirname);
    const { options: compilerOptions, errors } = typescript_1.default.convertCompilerOptionsFromJson(config.compilerOptions, dirname);
    if (errors && errors.length > 0) {
        throw errors;
    }
    const rootNames = await getRootNames(config, dirname);
    return { rootNames, compilerOptions };
}
exports.getProjectRootNamesAndCompilerOptions = getProjectRootNamesAndCompilerOptions;
function getTsConfigFilePath(project) {
    let configFilePath;
    let dirname;
    const projectStats = fs.statSync(project);
    if (projectStats.isDirectory()) {
        configFilePath = path.resolve(project, 'tsconfig.json');
        dirname = project;
    }
    else if (projectStats.isFile()) {
        configFilePath = project;
        dirname = path.dirname(project);
    }
    else {
        throw new Error("paramter 'project' should be a file or directory.");
    }
    return { configFilePath, dirname };
}
function getTsConfig(configFilePath, dirname) {
    const configResult = typescript_1.default.readConfigFile(configFilePath, p => fs.readFileSync(p).toString());
    const config = configResult.error ? {
        extends: undefined,
        compilerOptions: {
            lib: [
                'dom',
                'es5',
                'es2015',
                'es2016',
                'es2017'
            ],
            allowSyntheticDefaultImports: true
        }
    } : configResult.config;
    if (config.extends) {
        let project;
        if (path.isAbsolute(config.extends)) {
            project = config.extends;
        }
        else if (config.extends === '.'
            || config.extends === '..'
            || config.extends.startsWith(`.${path.sep}`)
            || config.extends.startsWith(`..${path.sep}`)
            || config.extends.startsWith('./')
            || config.extends.startsWith('../')) {
            project = path.resolve(dirname, config.extends);
        }
        else if (config.extends.endsWith('.json')) {
            project = path.resolve(dirname, 'node_modules', config.extends);
        }
        else {
            project = path.resolve(dirname, 'node_modules', config.extends + '.json');
        }
        const { configFilePath, dirname: extendsBasename } = getTsConfigFilePath(project);
        const extendsConfig = getTsConfig(configFilePath, extendsBasename);
        config.compilerOptions = { ...extendsConfig.compilerOptions, ...config.compilerOptions };
    }
    return config;
}
async function getRootNames(config, dirname) {
    const include = config.include;
    const exclude = config.exclude || ['node_modules/**'];
    if (config.files) {
        return config.files.map(f => path.resolve(dirname, f));
    }
    if (include && Array.isArray(include) && include.length > 0) {
        const rules = [];
        for (const file of include) {
            const currentPath = path.resolve(dirname, file);
            const stats = await statAsync(currentPath);
            if (stats === undefined) {
                rules.push(currentPath);
            }
            else if (stats.isDirectory()) {
                rules.push(`${currentPath.endsWith('/') ? currentPath.substring(0, currentPath.length - 1) : currentPath}/**/*.{ts,tsx}`);
            }
            else if (stats.isFile()) {
                rules.push(currentPath);
            }
        }
        return globAsync(rules.length === 1 ? rules[0] : `{${rules.join(',')}}`, exclude, dirname);
    }
    const rootNames = await globAsync(`**/*.{ts,tsx}`, exclude, dirname);
    return rootNames.map((r) => path.resolve(process.cwd(), dirname, r));
}
function statAsync(file) {
    return new Promise((resolve) => {
        fs.stat(file, (error, stats) => {
            if (error) {
                resolve(undefined);
            }
            else {
                resolve(stats);
            }
        });
    });
}
function globAsync(pattern, ignore, cwd) {
    return new Promise((resolve, reject) => {
        glob_1.default(pattern, { ignore, cwd }, (error, matches) => {
            if (error) {
                reject(error);
            }
            else {
                resolve(matches);
            }
        });
    });
}
